### PR TITLE
Display suggestions in Direction panel

### PR DESCRIPTION
**Pull request by G-Ray**
- Display suggestions inside DirectionPanel
- Use props `inputNode` (input to fetch suggests from) and `ouputNode` to display suggestions results. If no outputNode is provided, an absolute position will be computed to keep the existing behavior (like in top search bar)
- outputNode will be rendered thanks to `ReactDOM.createPortal`

### CODE REVIEW

File: src/panel/direction/DirectionForm.jsx [hunk 1]
```diff
@@ -56,7 +56,17 @@ export default class DirectionForm extends React.Component {
       destinationInputText,
     } = this.props;
 
+    const isMobile = isMobileDevice();
+
+    const VehicleSelectorComponent = props => <VehicleSelector
```
Comment (bbecquet) [inline src/panel/direction/DirectionForm.jsx:61]: I don't see the point of declaring a function here. Why not using `VehicleSelector` directly?

File: src/panel/direction/DirectionForm.jsx [hunk 2]
```diff
@@ -56,7 +56,17 @@ export default class DirectionForm extends React.Component {
...
+      vehicles={vehicles}
+      activeVehicle={activeVehicle}
+      onSelectVehicle={onSelectVehicle}
+      {...props}
+    />;
+
     return <div className="itinerary_form">
+      {!isMobile && <VehicleSelectorComponent />}
```
Comment (bbecquet) [inline src/panel/direction/DirectionForm.jsx:69]: It it's only a display order which changes, you can keep exactly the same DOM and use CSS.
For example, `{display: flex; flex-direction: column-reverse}` on the container will do the trick.

File: src/panel/direction/DirectionInput.jsx [hunk 3]
```diff
@@ -80,6 +81,7 @@ class DirectionInput extends React.Component {
   render() {
     const { pointType, inputRef, isLoading } = this.props;
     const { mounted, readOnly } = this.state;
+    const isMobile = isMobileDevice();
```
Comment (bbecquet) [inline src/panel/direction/DirectionInput.jsx:84]: Same component as before. For this to be reactive to screen size changes, you have to use the context, not the function directly. An example, on another class component https://github.com/QwantResearch/erdapfel/blob/master/src/panel/poi/PoiPanel.jsx#L332

File: src/components/ui/SuggestsDropdown.jsx [hunk 4]
```diff
@@ -5,26 +5,32 @@ import { object, func, string, arrayOf } from 'prop-types';
 import SuggestItem from './SuggestItem';
 
 const SuggestsDropdown = ({
+  inputNode,
+  outputNode,
```
[the remark this answers is not in the document]
Comment (G-Ray) [reply src/components/ui/SuggestsDropdown.jsx]: done :+1:

File: src/components/ui/SuggestsDropdown.jsx [hunk 5]
```diff
@@ -5,26 +5,32 @@ import { object, func, string, arrayOf } from 'prop-types';
...
+        width: boundingRect.width,
+      };
+    })();
 
   useEffect(() => {
     const keyDownHandler = e => {
       const { key } = e;
 
-      if (document.activeElement.getAttribute('id') !== inputId) {
+      if (document.activeElement.getAttribute('id') !== inputNode.id) {
```
[the remark this answers is not in the document]
Comment (G-Ray) [reply src/components/ui/SuggestsDropdown.jsx]: Yeah, I think it was needed when the suggest was mounted on search bar with ReactDom

File: src/panel/direction/DirectionForm.jsx (hunk 2 above)
Comment (G-Ray) [reply src/panel/direction/DirectionForm.jsx]: yep, done

File: src/components/ui/Suggest.jsx [hunk 6]
```diff
@@ -1,27 +1,40 @@
...
-      fetchItems(searchInputDomHandler.value);
+      if (inputNode.value === '') {
+        setIsOpen(true);
+        fetchItems('');
+      } else {
+        inputNode.select();
+        if (isMobile) {
+          setIsOpen(true);
+          fetchItems(inputNode.value);
+        }
```
Comment (amatissart) [inline src/components/ui/Suggest.jsx:36]: As a consequence, it's now impossible to see the suggestions list after the search field is blurred and no result has been selected. Is it really expected ?

File: src/components/ui/Suggest.jsx [hunk 7]
```diff
@@ -50,57 +63,74 @@ const Suggest = ({
     }, SUGGEST_DEBOUNCE_WAIT);
 
     const handleInput = e => {
-      fetchItems(e.target.value);
+      const { value } = e.target;
+      if (value !== lastQuery) {
+        fetchItems(value);
+        setIsOpen(true);
+      }
```
Comment (amatissart) [inline src/components/ui/Suggest.jsx:70]: The value of `lastQuery` is not updated here during typing: as far as I understand the component is not re-rendered and the updated state value is not accessible.

Actually I am not sure this check is useful. `fetchItems` is already responsible for caching results, so maybe we could remove this check completely.

File: src/panel/direction/DirectionPanel.jsx [hunk 8]
```diff
@@ -332,6 +332,7 @@ export default class DirectionPanel extends React.Component {
           className="direction_panel"
         >
           {form}
+          <div id="itinerary_autocomplete_suggestions" />
```
Comment (amatissart) [inline src/panel/direction/DirectionPanel.jsx:335]: The suggestions list remain visible after the last field (that triggers the directions search) is submitted with Enter key:

File: src/components/ui/Suggest.jsx (hunk 7 above)
Comment (G-Ray) [reply src/components/ui/Suggest.jsx:70]: I think you're right about removing this check. `lastQuery` was updated in `fetchItems`. I moved this update in this `handleInput` function, thanks. `lastQuery` is notably used to restore what the user has previously typed when cycling over suggests results with arrow keys. (after one complete cycle, the typed value is restored).

File: src/components/ui/Suggest.jsx (hunk 6 above)
Comment (G-Ray) [reply src/components/ui/Suggest.jsx:36]: you mean on desktop right ? This is the current design choice, we should contribute your comment to the discussion about design, what do you think ?

File: src/panel/direction/DirectionPanel.jsx (hunk 8 above)
Comment (amatissart) [reply src/panel/direction/DirectionPanel.jsx:335]: This bug turns out to be specific to Firefox.

Comment (G-Ray) [reply src/panel/direction/DirectionPanel.jsx:335]: thanks, it should be fixed now

File: src/panel/direction/DirectionInput.jsx [hunk 9]
```diff
@@ -23,11 +24,21 @@ class DirectionInput extends React.Component {
   }
 
   componentDidMount() {
+    if (this.props.isLoading) {
+      this.props.inputRef.current.blur();
```
Comment (G-Ray) [inline src/panel/direction/DirectionInput.jsx:28]: for instance, `isLoading` is true when the app is restoring origin and destination points from the url after a reload[...]

File: src/panel/direction/DirectionInput.jsx [hunk 10]
```diff
@@ -23,11 +24,21 @@ class DirectionInput extends React.Component {
...
+    }
+
     this.setState({
       mounted: true,
     });
   }
 
+  componentDidUpdate(prevProps) {
+    if (!prevProps.isLoading && this.props.isLoading) {
+      this.props.inputRef.current.blur();
```
Comment (G-Ray) [inline src/panel/direction/DirectionInput.jsx:38]: [...] Or when selecting a poi. This will implicitly close the dropdown.

**Pull request merged**